src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -0,0 +1,477 @@
+            let errorBytes = encoder.encode(error.message);
+            let written = errorBytes.length;
+            byteBuffer.set(errorBytes);
+            intBuffer[1] = written;
+            Atomics.store(intBuffer, 0, ERROR_EXCEPTION);
+            Atomics.notify(intBuffer, 0);
+        }
+    }
+});
+self.postMessage({inited:true})
+"""
+
+
+def _obj_from_dict(dict_val: dict[str, Any]) -> JsProxy:
+    return to_js(dict_val, dict_converter=js.Object.fromEntries)
+
+
+class _ReadStream(io.RawIOBase):
+    def __init__(
+        self,
+        int_buffer: JsArray,
+        byte_buffer: JsArray,
+        timeout: float,
+        worker: JsProxy,
+        connection_id: int,
+    ):
+        self.int_buffer = int_buffer
+        self.byte_buffer = byte_buffer
+        self.read_pos = 0
+        self.read_len = 0
+        self.connection_id = connection_id
+        self.worker = worker
+        self.timeout = int(1000 * timeout) if timeout > 0 else None
+        self.is_live = True
+
+    def __del__(self) -> None:
+        self.close()
+
+    def close(self) -> None:
+        if self.is_live:
+            self.worker.postMessage(_obj_from_dict({"close": self.connection_id}))
+            self.is_live = False
+        super().close()
+
+    def readable(self) -> bool:
+        return True
+
+    def writeable(self) -> bool:
+        return False
+
+    def seekable(self) -> bool:
+        return False
+
+    def readinto(self, byte_obj: Buffer) -> int:
+        if not self.int_buffer:
+            return 0
+        if self.read_len == 0:
+            # wait for the worker to send something
+            js.Atomics.store(self.int_buffer, 0, 0)
+            self.worker.postMessage(_obj_from_dict({"getMore": self.connection_id}))
+            if js.Atomics.wait(self.int_buffer, 0, 0, self.timeout) == "timed-out":
+                raise _TimeoutError
+            data_len = self.int_buffer[0]
+            if data_len > 0:
+                self.read_len = data_len
+                self.read_pos = 0
+            elif data_len == ERROR_EXCEPTION:
+                raise _StreamingError
+            else:
+                # EOF, free the buffers and return zero
+                self.read_len = 0
+                self.read_pos = 0
+                self.int_buffer = None
+                self.byte_buffer = None
+                return 0
+        # copy from int32array to python bytes
+        ret_length = min(self.read_len, len(memoryview(byte_obj)))
+        subarray = self.byte_buffer.subarray(
+            self.read_pos, self.read_pos + ret_length
+        ).to_py()
+        memoryview(byte_obj)[0:ret_length] = subarray
+        self.read_len -= ret_length
+        self.read_pos += ret_length
+        return ret_length
+
+
+class _StreamingFetcher:
+    def __init__(self) -> None:
+        # make web-worker and data buffer on startup
+        self.streaming_ready = False
+
+        dataBlob = js.Blob.new(
```

Comment:
I don't know if there are conventions for this already, but it might be nice to have JS magic objects named with a `js_` prefix or the like? So we know when we're passing things back and forth between Python and JS-land.

Reply:
good idea, is done.